### ISSUE TITLE
Library no longer working with latest Xbox / Xbox app updates

### ISSUE THREAD

**Issue by wh010ne** [OpenXbox/xbox-smartglass-csharp#64]
Is it me or did MS shit on us? Discovery no longer works, Authenticate doesn't work (except for Python version), Ping works still. Connect doesn't work even with tokens... Is anyone else having trouble?

**Comment by joelday** [comment]
I wouldn't equate breaking changes to a third party client with being "shit on".

**Comment by wh010ne** [comment]
Side note: Everything working fine until 5/23, when Xbox did a mandatory console update following the release of 3 zero-days from polarbear. Then things went south.

After working with it all night here are my observations:
   PingAsync - works fine
   DiscoverAsync - doens't work in CSharp (DiscoverAsync -a "ip address") works in Python only.
   Authenticate - works without redirect url (url MS returns is now invalid), it does generate tokens.json.
   Connect - works annoymously only, if I pass hash and and auth token - fails parsing the auth token 
         string (exception thrown on string length and character position) calling it as follows:

ConnectAsync(ipAddressOrHostname, "500000609750000000", "EwAIA+pvBAAUKods63Ys1fGlwiccIFJ+qE1hANsAAawrJWWvzqtuEwG71Qt6rU2rQfnyXFE2Ccz1q1dgumgEvjLjdrCKW8asgSK5HxB2DdhmVstk0X14szj/vwO9xWKluuLuFAujno4uiCR6DEgKFqEW5s04/RQkyUu63qsKdDeMdILdARhlvxUZLgBLxGTQIsO7g3bc5rNaUxAjAF1MMHZEEmlBwvxZWpsUBDKjeOwYl/5tK1I+ZJ/PsRXJWUUXSyQ5pwmZFjYNz8lM5p3iMGwoDW70mz9KmtsFE8p22JkNrl5rc0sn46tQhaNhdI6OkagdEx6O1KWwzkr6m4a3slumnSVIZqnp4Yl9hLReSEgrg7fZWsfuE603M9/jHTQDZgAACNxo3H2R3d8i2AEVmbv3SFjAnluGIIrDMd9nt58mukiPJwHF5ER+y7Abj2tPwN/wEHUiVlsqXhUznlgGkQhyXDn+ZqHqOhGwrstW4F9DtALVhLeGxPA8OTmWvu3Gqud+VO669Tg1LRMyVo01hBZxb6fw9vc6qO681hybmswSX8VPPE1aDxg36SG15UV7lFXFiekclk7+5FuzIlq8JqsJxzQIJfhSUqqjIo2MbQx8HS0rCRTjgv+x0ptda+mcjBrsgRdapt1iQWgTkqTBQHcAAmdDVB67ykM7CAusHyKf6G9mYFSXgd8KOtIeGGMZoJHNWhH29SaPfjKAeI2j2/jtDd/tJhdwCrH89Tx0Dcm934F45ZHRBO/aoTZnz/uTnmF7EOsVAj74mJAl+lldANhWMczS2t2HTspl+IPsYrcyQTxHzqM8nzGOhqSwN5Gy54Niawe8C9hvUT8/VqrT+LUoGyhSIErARbY9L48rcM37XohtOiJM8zi4IEivLBKIkCMjAsQEU9NNL9wWVKo8m6tLY/CKfQjycoqvfNZikeyWQ7YxBnTnaojSkq3AUckKCWRHTfnwBGPxRV2/BtjKk0GwfPcPlgibvnOpn4/DXUMlFUyU7bWzdDTJRA+kElaoRiT3d0iNHgI=");
   
Note: I've tried all tokens. All fail. I even tried using the shared secret dumped from Frida.
I can connect annonymously like ConnectAsync("ip") and start getting AWK's but cannot send any commands. I tried sending simple button presses and no response at all. The library does generate some exceptions through the output window but I'm still tracking those down, it not in my code.

Your probably right about equating 3rd party breaks not being "shit on" I guess in all fairness to evolving technology things are bound to change and we need to adopt, it just seems everytime users start getting working properly and how the like it, wham we're hit with breaking changes and it's very frustrating having to rework everything.

**Comment by tuxuser** [comment]
You are passing wrong authentication data to Connect. The userhash and token strings look very odd!

bogus values:
Userhash: "2347453934346"
Token: "eyWrekoskogkdfgdfgdfgnblabla..." // note it starts with "ey"

It looks like you are passing AccessToken to Connect.. which is wrong.
You want to use XToken parameters!

A note about SmartGlass.Cli -> Authenticate
There is a problem when pasting the redirect uri -> CMD.EXE and PowerShell have an input limit, it wont take the full string... on Unix shell its good.
You might want to do the authentication manually, in your code - instead of using the Cli.

**Comment by wh010ne** [comment]
Thnx tux! I will try that. Perhaps, the Gamepad state wasn't setting was because the token. Good to know that it IS working for someone though. I will work through it. Would that XToken be the same as XSTSToken? I've only got two tokens that start with "ey", UserToken and XSTSoken from tokens.json, or am I completely missing somewhere else to get the XToken?

**Comment by tuxuser** [comment]
XSTSToken (term I took from the old python xbox lib) == XToken (Actual name of the token, in Microsoft xbox live api)
It's the same thing ;)

**Comment by wh010ne** [comment]
Awesome thanks! I had a suspicion. I somehow recall XSTSToken from something to do with the 360. I've been disassembling the Durango SDK SmartGlass stuff, but I have a feeling that they've reworked the whole thing as now SmartGlass isn't supported on 360 anymore.

For anyone having allow Anonymous connections issues:
One thing I specifically had to do to get it to connect is on the console, I had "Allow all devices" to connect, but it wasn't accepting it until, I went into the console's advanced network settings and clear the alternate MAC, and restarted console, magically it now changed it's flags and let me connect. This didn't have anything to do with this library though, just a quirk with Xbox.

**Comment by wh010ne** [comment]
Interesting.. I moved everything over to Linux and it works perfectly. Basically all them same code (a .Net console app, referencing this package, and doing just Discover, Ping, SendGamePad, and Connect.) Works perfect on Linux, but Windows 10 1903h has weird issues. It won't even discover the console even with IP Address specified. Same code in Linux discovered the console within seconds. Mind you, this Linux instance is a VM sitting on my current dev OS (win10 1903, that doesn't work).  The only thing I can think of that I multi-targeted this package to Standard2.0 and Net4.6.2. Other than that I am at a loss, figuring out what it might be. The main thing is the Discover. I've tried playing with the message settings and nothing. I just never get any received messages responses under Windows. Going to recompile with only Standard2.0 and test for now.

 Thnx Tux and Joel for your feedback.

**Comment by wh010ne** [comment]
Note: WSL Linux (kali) doesn't work. Half the framework is very 'glitchy'. So, there's something with Windows 1903 causing issues on networking functionality. (no firewall running, no AV running, UPnP enabled, basically my dev machine is DMZ). WSL showed promise at first, but it failed Discovery also.

Linux VM x64 (mint 19.1) works perfectly.

P.S. Joel dont seem so butt hurt! I'm just asking for feedback, not insulting your creation. I realize things change and we adapt. np. You still created something other didn't.

**Comment by wh010ne** [comment]
Glad this is leaving a record. This thread will help future developers in similar situations. The Docs are good but difficult to interpret for examples.

Tux, Joel have you experimented with DevicePicker API yet? I played with it and it picked up the console and paired in seconds and adds a little more functionality to Device. I think I will try to cross host the two API's and test. Just not sure how to cast the Device from this package to DevicePicker.Device yet... arhh.

**Comment by wh010ne** [comment]
I guess this would be to you Joel, why doesn't discovery work on win (1903)? I get your prerequisites but it ain't working. Please post any updates.

**Comment by wh010ne** [comment]
Ok, before I close this I want to comment on final solution.

My app consists of a .Net 4.6.2 Console Application referencing xbox-smartglass-csharp (only SmartGlass project) as a included project in my solution 
![Untitled](https://user-images.githubusercontent.com/18432559/58374852-57b64e80-7efa-11e9-85f6-02553cf511ab.png)

I had SmartGlass targeting both .NetStandard2.0.3 and .Net4.6.2, while XboxController was targeting .Net4.6.2. Compiled fine, just wouldn't work with all functionality, especially Discovery.

SOLUTION:
Targeted SmartGlass to only .NetStandard2.0.3
Targeted XboxController to only .Net4.7.2

Compiled but now got exceptions when calling Discover. Specifically in AttributeMapping class of SmartGlass, complaining about 'missing/cannot load types' found out it was looking for NewtonSoft.Json even though it was already referenced via Nuget, but not in XboxController so I added it to XboxController also and it stopped complaining. So I referenced the same Nuget packages that exist in SmartGlass in XboxController also. Cleaned solution and rebuilt, suddenly Discovery sprang to life.

The library is now working as it was intended. Hope this helps others in similar situations.
